crates/sources/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -1,57 +1,237 @@
+                },
+            );
+        }
+
         Ok(Self {
-            registered_pools: graph_api.get_registered_pools().await?,
+            pools_by_token_pair,
             graph_api,
+            cache: Mutex::new(pools),
+            max_age,
         })
     }
+
+    /// Assumes that all pool ids already exist in the cache
+    async fn get_pools_and_update_cache(&self, pool_ids: &[H160]) -> Result<Vec<PoolData>> {
+        let pools = self.graph_api.get_pools_with_ticks_by_ids(pool_ids).await?;
+        let now = Instant::now();
+        let mut cache = self.cache.lock().unwrap();
+        for pool in &pools {
+            cache.insert(
+                pool.id,
+                CachedPool {
+                    pool: pool.clone(),
+                    updated_at: now,
+                    requested_at: now,
+                },
+            );
+        }
+        Ok(pools)
+    }
+
+    /// Returns cached pools and ids of outdated pools.
+    fn get_cached_pools(&self, token_pairs: &HashSet<TokenPair>) -> (Vec<PoolData>, Vec<H160>) {
+        let now = Instant::now();
+        let mut cache = self.cache.lock().unwrap();
+        token_pairs
+            .iter()
+            .filter_map(|pair| self.pools_by_token_pair.get(pair))
+            .flatten()
+            .partition_map(|pool_id| match cache.get_mut(pool_id) {
+                Some(entry) if now.saturating_duration_since(entry.updated_at) < self.max_age => {
+                    entry.requested_at = now;
+                    Either::Left(entry.pool.clone())
+                }
+                _ => Either::Right(pool_id),
+            })
+    }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for UniswapV3PoolFetcher {
-    async fn fetch(&self, _token_pairs: HashSet<TokenPair>, _at_block: Block) -> Result<Vec<Pool>> {
-        todo!()
+    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolData>> {
+        let (mut cached_pools, outdated_pools) = self.get_cached_pools(token_pairs);
+
+        if !outdated_pools.is_empty() {
+            let updated_pools = self.get_pools_and_update_cache(&outdated_pools).await?;
+            cached_pools.extend(updated_pools);
+        }
+
+        Ok(cached_pools)
+    }
+}
+
+pub struct CachingUniswapV3PoolFetcher(Arc<UniswapV3PoolFetcher>);
+
+impl CachingUniswapV3PoolFetcher {
+    /// Creates new CachingUniswapV3PoolFetcher with the purpose of spawning an additional
+    /// background task for periodic update of cache
+    pub async fn new(chain_id: u64, max_age: Duration, client: Client) -> Result<Self> {
+        Ok(Self(Arc::new(
+            UniswapV3PoolFetcher::new(chain_id, max_age, client).await?,
+        )))
+    }
+
+    /// Spawns a background task maintaining the cache once per `update_interval`.
+    /// Only soon to be outdated prices get updated and recently used prices have a higher priority.
+    /// If `update_size` is `Some(n)` at most `n` prices get updated per interval.
```

Comment:
I think this could be done, but let's add it later as an improvement, so we don't increase the complexity of this PR.
(Currently new indexed pools will be added on pod restart which is not as bad I guess)

But in order to do, I think another maintenance task should be added (that would be ran each, let's say, 1h), and it would add new values to the `self.pools_by_token_pair`, meaning the field should go under the mutex also.